src/boot/mexpr.ml

```diff
@@ -438,28 +448,28 @@ let rec val_equal v1 v2 =
   | _ -> false
 
 
-(* Convert a term into de Bruijn indices. *)
-let rec debruijn env t =
-  let rec find fi env n x =
-    (match env with
-     | VarTm(y)::ee -> if y =. x then n else find fi ee (n+1) x
-     | [] -> raise_error fi ("Unknown variable '" ^ Ustring.to_utf8 x ^ "'")) in
+(* Add symbol associations between lambdas, patterns, and variables *)
+let rec symbolize env t =
+  let findsym fi x kind env = try List.assoc x env
+    with Not_found -> raise_error fi ("Unknown " ^ kind ^ " '" ^ Ustring.to_utf8 x ^ "'")
+  in
   let rec dbPat env = function
-    | PatNamed(_,NameStr(x)) as pat -> (VarTm(x)::env, pat)
+    | PatNamed(fi,NameStr(x,_)) -> let s = gensym() in ((x,s)::env, PatNamed(fi,NameStr(x,s)))
     | PatNamed(_,NameWildcard) as pat -> (env, pat)
     | PatSeq(fi,ps,seqMP) ->
        let go p (env,ps) = let (env,p) = dbPat env p in (env,Mseq.cons p ps) in
        let (env,ps) = Mseq.fold_right go ps (env,Mseq.empty) in
-       let env = (match seqMP with
-       | SeqMatchPrefix(NameStr(x)) | SeqMatchPostfix(NameStr(x)) -> VarTm(x)::env
-       | SeqMatchPrefix(NameWildcard) | SeqMatchPostfix(NameWildcard) | SeqMatchTotal -> env) in
-       (env,PatSeq(fi,ps,seqMP))
+       let (env,seqMP') = (match seqMP with
+         | SeqMatchPrefix(NameStr(x,_)) -> let s = gensym() in ((x,s)::env, SeqMatchPrefix(NameStr(x,s)))
+         | SeqMatchPostfix(NameStr(x,_)) -> let s = gensym() in ((x,s)::env, SeqMatchPostfix(NameStr(x,s)))
+         | SeqMatchPrefix(NameWildcard) | SeqMatchPostfix(NameWildcard) | SeqMatchTotal -> (env,seqMP)) in
+       (env,PatSeq(fi,ps,seqMP'))
     | PatTuple(fi,ps) -> (* NOTE: this causes patterns to introduce names right-to-left, which will cause errors if a later pattern binds a name that is seen as a constructor in a pattern to the left *)
        let go p (env,ps) = let (env,p) = dbPat env p in (env,p::ps) in
        let (env,ps) = List.fold_right go ps (env,[])
        in (env,PatTuple(fi,ps))
     | PatCon(fi,cx,_,p) ->
-       let cxId = find fi env 0 cx in
+       let cxId = findsym fi cx "constructor" env in
```

Comment:
This means that constructors and normal variables still live in the same namespace, which I believe isn't what we want in the long run (it isn't quite the case in the interpreter in mlang). Not sure if we should fix that now or later.

Reply:
I was planning to rework some parts on constructors in another patch.